### ISSUE TITLE
Train new base network. Assert performance. 

### ISSUE THREAD

**Issue by beijbom** [beijbom/coralnet#234]
This is a huge issue, which is detailed further in
https://docs.google.com/document/d/1mL2n6KFi2CAcTpMI3OXBnHFIkkfVWBDEtDmGGLRU4hs/edit

**Comment by beijbom** [comment]
@qiminchen @kriegman : what is the latest estimate for when this ticket is all the way done? In other words: how is the production training going?

**Comment by qiminchen** [comment]
@kriegman @beijbom Oscar, I just had a meeting with David today, here is the summary of all the works that I've been doing recently. Please refer to [Sheet](https://docs.google.com/spreadsheets/d/15NxI4AYrADqBB6KBQ-sL_ruEEf8mT10j_bq4w0tVhqI/edit?usp=sharing) for all quantitative results.

**1. Production training is done**

Training setup: EfficientNet - b0, Adam Optimizer, w/o augmentation, manually adjust the learning rate

- sheet `v2/v3/v4`, column `D-F`: Accuracy of each test source and overall accuracy in the last row, along with the accuracy improvement compared with `v1` and `backend`.
- sheet `v4 sources details`, column `J-L`: `precision`, `recall` and `F1-score` of each test source.
- sheet `v4 labels details`, column `M-O`: `precision`, `recall` and `F1-score` of each label in the test sources, scroll down to the last row for the overall weighted performance.

I am gonna start production training on `ResNet101` and see how far it can go.


**2. Other ML classification models**

Besides the `Logistic regression` classifier, I also tried `Logistic regression with class_weight`, `Linear SVM`, `Random Forest`, and `Gaussian Naive Bayes`. Full quantitative results in sheets.

- Logistic regression with `class_weight`

`SGDClassifier.partial_fit()` does not support [`class_weight`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_class_weight.html) but [`sample_weight`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_sample_weight.html), and `sample_weight` is directly computed by `class_weight` therefore is applied during training. This method does not perform as good as directly applying geometric augmentation but faster.

- Linear SVM

Changing the `SGDClassifier(loss='log', average=True)` to `SGDClassifier(loss='hinge', average=True)` gives us a Linear SVM model, performs slightly better than logistic regression classifier but takes a bit longer to train.

- [Random Forest](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

```
elif clf_method == 'rf':
    clf = RandomForestClassifier(warm_start=True, n_estimators=30, n_jobs=-1)
    with open(os.path.join('/mnt/sda/features/gamma/effnetb0_rf_4eps',
                           source, 'i000000.features.json'), 'r') as f:
        xrf = json.load(f)
    yrf = list(np.load(os.path.join('/mnt/sda/features/gamma/effnetb0_rf_4eps',
                                    source, 'i000000.features.anns.npy')))
    yrf = [classes_dict[i] for i in yrf]
```
This is a bit tricky since the `RandomForestClassifier` does not support `partial_fit()` so `warm_start=True` to reuse the solution of the previous call to fit and add more estimators to the ensemble. Also, it requires all labels to be presented in each chunk of data so I ended up appending each chunk of data with a list of features that include all the labels of that source.

`n_estimators`: tuned with `10`, `20`, `30`, `50` and `100`, the performance didn't change a lot when `n_estimators > 30`, but it takes way much longer to train when using `n_estimators=100`.

`max_features`: tuned with `sqrt(n_features)`, `log2(n_features)`, and `n_features`, ended up using `sqrt(n_features)` because using all features only improves the performance marginally but takes way much longer to train.

- [Gaussian Naive Bayes](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html)

Performance is trivial...

Let me know if you have any questions and Oscar, the production net weight is ready.

**Comment by kriegman** [comment]
One comment to add

class_weight was done to balance the number of samples per class. This
"upweights" the rare classes/downweights the common clases so their loss
per class are equally weighted.
This should be contrasted with our data augmentation strategy where we
augmented the rare classes and subsampled the larger classes, but didn't go
as far as "full balancing across classes."

Not revealed through the statistics is the impact on classification of
rarer classes as seen in through anecdotal exploration of the confusion
matrices. Basically, augmentation increased accuracy of rarer classes, but
at the price of lower accuracy for more prevalent classes. So, it wasn't an
unqualified win, but rather a tradeoff. Balancing wasn't as a effective as
augmentation at benefiting the rarer classes.

**Comment by beijbom** [comment]
Thanks @qiminchen . 0.37% boost compared to v1 isn't amazing, but you have done a great work building up to this and I don't think there are any obvious stones unturned. Looking forward to seeing what ResNet101 will do.

Can you share the training-loss curve? How many epochs did you end up training for?

Also: did you end up not using any data-augmentation at all? I thought we agreed on using `V4: 2-1275` which has class-balance + DA?

I'd suggest you go ahead and replace the EfficientNet_b0 version we have in spacer with this one. Perhaps we can simply update the weight and voila -- we have a new version! (This is OK at this point since this hasn't been deployed yet). You could also rename the extractor it but I think `efficientnet_b0_ver1` is as good of a name as any.

**Comment by beijbom** [comment]
... In fact, after you update the weights, can you add a check-sum to the `load_model` method to verify that we are actually downloading and using the correct weights? If we don't have that, we could accidentally over-write the weights-file on s3 and suddenly have a different feature extractor without knowing it. Ugh.

**Comment by qiminchen** [comment]
We were hovering between `V4: 1-1275` and `V4: 2-1275` and I ended up using `V4: 1-1275` setting because I didn't see many improvements but it takes longer for training with geometric augmentation.

As for the training, I ended up training the net with 22 epochs with a batch size of 156, and I logged the batch loss every 10,000 batches. I should've used tensorboard instead... The current best model was from epoch 21st.

![production](https://user-images.githubusercontent.com/45860234/84125283-30d99600-a9f1-11ea-8bc0-045b5cd874cc.png)

**Comment by beijbom** [comment]
Ok, thanks for sharing. I think those look good. I would probably have
deferred the first drop a bit later, but I don't think it makes a big
difference.

**Comment by qiminchen** [comment]
@kriegman @beijbom hey guys, here are some updates about the performance of `Logistic Regression` and `Multi-Layer Perceptron`, [Google Sheet](https://docs.google.com/spreadsheets/d/15NxI4AYrADqBB6KBQ-sL_ruEEf8mT10j_bq4w0tVhqI/edit?usp=sharing) for full quantitative results.

**TL;DR**

`Multi-Layer Perceptron` optimal parameters setting: 
- `hidden_layer_size = (200,100)` for `# of total training patch > 10,000` and `hidden_layer_size = (200)` for `# of total training patch <= 10,000`
- `learning_rate=1e-4`, 
- `optimizer=Adam`

1. `Multi-Layer Perceptron` performs better than `Logistic Regression` in both `recall` and `F1-score` with a considerable margin.
2. `Multi-Layer Perceptron` is faster than `Logistic Regression` during training.
3. `Multi-Layer Perceptron` performs better than `Logistic Regression` regardless of the relative source size: `# of total training patch / # of class`.
4. `ResNet101` training is at epoch 6 now.

`== >` We may consider replacing `Logistic Regression` with `Multi-Layer Perceptron`.

**Details**

1. `Multi-Layer Perceptron` hyperparameters tuning

I've tried different setups of `hidden_layer_size` and `learning_rate` in `5` test sources with different source size (`# of total training patches`). `s16` and `s1420` have the largest and smallest source size respectively. By looking through different setups over `5` sources, it's pretty obvious:
- `1e-3` is too large and `1e-5` is too small
- `(200, 100) or (200)` is preferable, no need to use more than 2 hidden layers

| source | s16 | s294 | s367 | s672 | s1420 |
| :-------: | :---: |  :---: |  :---: |  :---: |  :---: | 
| size | 358,272 | 34,906 | 13,435 | 8,028 | 5,100 |

- `s16` - with a size of `358,272`

*Note*: plots grouped by `hidden_layer_size` and `learning_rate` respectively for detailed comparison.

![s16_hls](https://user-images.githubusercontent.com/45860234/84842200-6128c200-aff9-11ea-8123-8ada25037fd5.png)
![s16_lr](https://user-images.githubusercontent.com/45860234/84842232-7271ce80-aff9-11ea-9d8b-58d528281e6b.png)
- `s294` - with a size of `34,906`

![s294_hls](https://user-images.githubusercontent.com/45860234/84842351-b95fc400-aff9-11ea-8db0-d184afffa119.png)
![s294_lr](https://user-images.githubusercontent.com/45860234/84842357-bbc21e00-aff9-11ea-8b05-644a9ad399b3.png)
- `s367` - with a size of `13,435`

![s367_hls](https://user-images.githubusercontent.com/45860234/84843065-c5e51c00-affb-11ea-91df-60821c708e7b.png)
![s367_lr](https://user-images.githubusercontent.com/45860234/84843066-c7164900-affb-11ea-84c9-9336175db46c.png)
- `s672` - with a size of `8,028`

![s672_hls](https://user-images.githubusercontent.com/45860234/84843125-e9a86200-affb-11ea-8b0c-6ef5965ec9c8.png)
![s672_lr](https://user-images.githubusercontent.com/45860234/84843130-ed3be900-affb-11ea-965f-bfbea8a16200.png)
- `s1420` - with a size of `5,100`

![s1420_hls](https://user-images.githubusercontent.com/45860234/84843421-b0bcbd00-affc-11ea-8614-ebe1fa4f8d1d.png)
![s1420_lr](https://user-images.githubusercontent.com/45860234/84843425-b1edea00-affc-11ea-975d-37e7d95fbe15.png)

2. Training time

`Multi-Layer Perceptron` is faster than `Logistic Regression` in most of the test sources. No tradeoff happened in training time.

![lr_mlp_runtime](https://user-images.githubusercontent.com/45860234/84844870-37bf6480-b000-11ea-820b-0ed8877d7692.png)

3. Accuracy of `26` test sources

- x-axis: the relative source size (`# of total training patch / # of class`), left to right -> small to large
- y-axis: the accuracy

![lr_mlp_acc](https://user-images.githubusercontent.com/45860234/84844991-7fde8700-b000-11ea-9b26-0f30348e81ae.png)

**Comment by kriegman** [comment]
So, if I'm reading the spreadsheet correctly, the most important numbers
are:

CoralNet Beta (backend) accuracy: 71.36
New Production (Efficient Net B0)  with LR classifier accuracy: 76.57
New Production (Efficient Net B9)  with MLP classifier accuracy: 76.82

CoralNet Beta (backend) F1: ?? [Not in table]
New Production (Efficient Net B0)  with LR classifier F1: 75.02
New Production (Efficient Net B9)  with MLP classifier F1: 76.37

So, the improvement for accuracy is modest (0.25%), but is more significant
for F1 (1.35%).

But I don't understand your statement: "Multi-Layer Perceptron performs
better than Logistic Regression regardless of the relative source size: #
of total training patch / # of class."

The last bar plot shows LR being more accurate for some sources and MLP
more accurate for other sources. Am I missing something?

**Comment by qiminchen** [comment]
>But I don't understand your statement: "Multi-Layer Perceptron performs
better than Logistic Regression regardless of the relative source size: #
of total training patch / # of class."
The last bar plot shows LR being more accurate for some sources and MLP
more accurate for other sources. Am I missing something?

You are right, `Multi-Layer Perceptron` performs better than `Logistic Regression` in most of the test sources, for `s609`, `s843`, `s1579`, `1559`, `s1070`, `Multi-Layer Perceptron` is slightly worse.

**Comment by qiminchen** [comment]
@kriegman 
>But I don't understand your statement: "Multi-Layer Perceptron performs
better than Logistic Regression regardless of the relative source size: #
of total training patch / # of class."
The last bar plot shows LR being more accurate for some sources and MLP
more accurate for other sources. Am I missing something?

You are right, `Multi-Layer Perceptron` performs better than `Logistic Regression` in most of the test sources, for `s609`, `s843`, `s1579`, `1559`, `s1070`, `Multi-Layer Perceptron` is slightly worse.

**Comment by beijbom** [comment]
@qiminchen : what is the status on this ticket? Any outstanding work-items? Did resnet 101 finish? Did you want to add MLP to the spacer codebase? Did you end up writing up a report on your work? CC @kriegman .

**Comment by qiminchen** [comment]
yes, ResNet101 pretrained on V1 and trained for another 8 epochs works better than EfficientNetb0 as you could see in [Sheet](https://docs.google.com/spreadsheets/d/15NxI4AYrADqBB6KBQ-sL_ruEEf8mT10j_bq4w0tVhqI/edit?usp=sharing) `EfficiNetb0 vs. ResNet101`. For a fair comparison, I also train EfficientNetbo pretrained on V1. The result will be available soon. And I will write up a report on EfficientNetb0 and ResNet101 on Logistic Regression and Multi-layer Perceptron very soon.

**Comment by beijbom** [comment]
Okay, thanks. What is your timeline for the report? We will need to make a decision which one to use on the production server.

**Comment by qiminchen** [comment]
By Tues, EfficientNetbo pretrained on V1 would be available by then

**Comment by kriegman** [comment]
The short answer from last monday was that ResNet50 had an absolute error
that was about 2% lower than EfficientNetB0.
But ResNet 50 is about half the speed and comparable to VGG16.

A month ago or so, Qimin was unsuccessful at training ResNet50 from
scratch, and so I asked him to initialize from 1-cycle optimizer version,
and that gave the boost.
So, he's gone back to EfficientNetB0 with the same strategy.

So, we'll need to make this decision of accuracy vs. speed.  [Hint; I'm
favoring accuracy].

Let's get a full report on this, make a decision, and get ready to deploy
it, presumably with MLP. But again, no commitment until we see the report.

D

**Comment by beijbom** [comment]
@qiminchen : I'll close this issue now since the work is done.
